crates/datasources/src/sqlite/wrapper.rs

```diff
@@ -0,0 +1,184 @@
+                            })
+                            .collect::<Result<Vec<_>, rusqlite::Error>>()
+                    });
+
+                    loop {
+                        // Collect the next "n" rows for the batch.
+                        const BATCH_SIZE: usize = 1000;
+
+                        let data = (0..BATCH_SIZE)
+                            .filter_map(|_| rows.next())
+                            .collect::<Result<Vec<_>, _>>();
+
+                        let batch = match data {
+                            Ok(data) if data.is_empty() => {
+                                // Exit when there's nothing more to process. This
+                                // should drop the only response sender and hence,
+                                // end the stream.
+                                break;
+                            }
+                            Ok(data) => Ok(SqliteBatch {
+                                cols: cols.clone(),
+                                data,
+                            }),
+                            Err(e) => Err(SqliteError::from(e)),
+                        };
+
+                        if tx.blocking_send(batch).is_err() {
+                            // Receiver is dropped so we can exit.
+                            break;
+                        }
+                    }
+
+                    Ok(())
+                })
+                .await?;
+            Ok(())
+        });
+
+        SqliteQueryStream {
+            rx,
+            _handle: handle,
+        }
+    }
+
+    // Collects and returns all the rows from the query.
+    pub async fn query_all(&self, s: impl Into<String>) -> Result<SqliteBatch> {
+        let s = s.into();
+
+        Ok(self
+            .inner
+            .conn(move |conn| {
+                let mut stmt = conn.prepare(&s)?;
+
+                let cols = stmt
+                    .columns()
+                    .into_iter()
+                    .map(|c| Column {
+                        name: c.name().to_owned(),
+                    })
+                    .collect::<Vec<_>>();
+
+                let num_cols = cols.len();
+
+                let data = stmt
+                    .query([])?
+                    .mapped(|r| {
+                        (0..num_cols)
+                            .map(|idx| {
+                                let v = r.get_ref(idx)?;
+                                Ok(Value::from(v))
+                            })
+                            .collect::<Result<Vec<_>, rusqlite::Error>>()
+                    })
+                    .collect::<Result<Vec<_>, rusqlite::Error>>()?;
+
+                Ok(SqliteBatch {
+                    cols: cols.clone(),
+                    data,
+                })
+            })
+            .await?)
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Column {
+    pub name: String,
+}
+
+#[derive(Debug, Clone)]
+pub struct SqliteBatch {
```

Comment:
I'd be inclined to avoid writing to this structure and just construct/build record batches directly.

Reply:
Sure, we can do that! Not very keen on doing it for `query_all` function though. Makes it hard to read values from record batch (for eg, when trying to query schema)

Reply:
On a second thought, this does make the code very hard to convert.

We will need an intermediate vector of rows and this serves as that. We'll have to maintain all fields as a vector of array builders (which do not implement any common trait).

```
// No such trait as an ArrayBuilder...
let fields: Vec<_> = schema.fields().map(|f| Arc::new(ArrayBuilder::from(f.data_type())).collect();
```

Even if they do (we could `Any` here for the reference):

```
let fields: Vec<Arc<dyn Any>> = schema.fields().map(|f| Arc::new(f.data_type().array_builder())).collect();
```

We'd have to downcast for each column. Not sure if it's worth it.

I'll push the changes in a commit, let's see if it feels OK.

Reply:
we did this (roughly) with the bson converter, maintained a vector of arrays for each column, and while you need an array for the _current_ column, you don't need _all_ of the rows. 

This avoids ever (basically) having more than (a record batch size)-number of rows in memory at the time.

Reply:
ohh okay! so i never looked it up but there is such a thing as `ArrayBuilder`! Should be easy then

Reply:
Made this change.